VACMS-19681: Event Form Conditionals Tests

chri5tia opened the pull request:
## Description

~Closes #10605, #10681, #10682~  Listing them like this has No magic and only connected 10605
Listing them each on their own line, connected all three issues to this PR.
Closes #10605
Closes #10681
Closes #10682

## Testing done
- [ ] Run cypress for local testing: `npm run test:cypress:interactive`, then select `content_editing.feature`
- [ ] Watch for a successful test message
- [x] Go over the Scenario and spec text to see if any logic is missed
- [ ] I also isolated just the tests I wrote and ran them by themselves (Not necessary for QA)

## Screenshots
<img width="903" alt="Screen Shot 2022-11-16 at 3 06 00 PM" src="https://user-images.githubusercontent.com/22764938/202315674-cac13e51-aed2-4a6b-98ce-fecab9d757f2.png">

## QA steps

- [ ] Pull this project branch
- [ ] See the "Testing done" section and repeat it.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`


### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

chri5tia commented:
Note: Some cypress tests are failing on `content_editor.feature`, but they are also failing on the main branch that does _not_ contain my changes. I will need to cut a ticket to evaluate these pre-existing failures. Most likely changes were made to node forms and the tests need to be updated to reflect that.
<img width="889" alt="Screen Shot 2022-11-16 at 4 28 31 PM" src="https://user-images.githubusercontent.com/22764938/202324523-73815260-708c-43af-8a99-c8788b29e407.png">

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I should see "State"
+    And I should see "Country"
+    And I select the "Online" radio button
+    Then I should see an element with the selector "#edit-field-url-of-an-online-event-0-uri"
```
I like the way this one uses a field id as it make it less brittle.  The others using field labels is easier to read, but more likely to fail with a label change.

swirtSJW replied:
On second thought  The readability is preferred.  If the label changes, the person changing it can fix the failing test.

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    # Location type selector reveals conditional form elements
+    And I select the "At a VA facility" radio button
+    Then I should see "Facility location"
+    And I should see "Building, floor, or room"
```
I think this is only showing half the spec (desired behavior).  The spec is that some should show AND other should hide.  Right now this is only testing the the showing.  However, if something broke our hiding js and all the fields were showing, this test would still pass.

I think you need to add  a bunch of "And I should not see"s to actually test both sides of the desired behavior.

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    # Registration checkbox reveals conditional form elements
+    And I check the "Include registration information" checkbox
+    Then I should see "Registration is required for this event"
+    And I should see "Call to action"
```
Similar concern here.  We are only testing the showing, but not the hiding.  IF the JS failed completely, there would be no hiding, but this test would pass.

chri5tia replied:
This one is easy because everything shows except for in one instance one field doesn't and I covered that.

swirtSJW replied:
I think you misunderstood my point  You are only looking for what becomes visible when "I check registration information" checkbox.

it was never evaluated that the fields you expecting to see, were ever hidden.  So only one state is being tested.

I would expect that there was an initial check to make sure  cost, registration is required and call to action were hidden initially, AND OR  hide "When I uncheck the "Include registration information" checkbox"

chri5tia replied:
Well now I'm seeing an issue with the use of "I should not see" here because the current definitions are that "see/not see" relate to existence and not visibility still. It will be nice to get that sorted out. For now, I'll need to change all of those to "should not be visible" and "should be visible" to ensure that the test ins't just making sure that stuff exists. (I think. I thought I fixed this?)

ndouglas replied:
Yes, let's just use "should (not) be visible" right now and then we'll straighten out the rest in a separate PR.

swirtSJW reviewed (changes requested):
This is great work @chri5tia  I only see one problem with the conditional display checks.  See my comments below.

swirtSJW commented:
> Most likely changes were made to node forms and the tests need to be updated to reflect that.

No.  The tests would prevent that scenario because any pr that changed the node form would not get passing tests so would not get merged.  These failures are transient.

swirtSJW commented:
I edited the Closes #xxxx, #xxy, #xyx
to be on separate lines so that all three issues get connected
![image](https://user-images.githubusercontent.com/5752113/202350099-4ba15bdb-4c1c-4f8a-b33f-f6918eafe948.png)

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I am at "node/add/event"
+    And the element with selector "#edit-field-datetime-range-timezone-0-timezone" should contain "New York"
+
+  Scenario: Confirm that the event form conditional elements are show or hidden appropriately
```
```suggestion
  Scenario: Confirm that the event form conditional elements are shown or hidden appropriately
```

ndouglas commented on tests/cypress/integration/behavioral/common/should_be_visible.js:
```diff
@@ -0,0 +1,27 @@
+import { Then } from 'cypress-cucumber-preprocessor/steps';
+
+// TODO: Remove this step definition and change "I should see" to accomplish this
+//       Then create "Should not exist" to use .should('exist') instead of "I should see"
+//       See VACMS-11609
```
👍🏻

ndouglas commented on tests/cypress/integration/behavioral/common/should_be_empty.js:
```diff
@@ -0,0 +1,4 @@
+import { Then } from 'cypress-cucumber-preprocessor/steps';
+
+Then('the element with selector {string} should be empty', (selector) => cy.get(selector).should('be.empty'));
+Then('the element with selector {string} should not be empty', (selector) => cy.get(selector).should('not.be.empty'));
```
```suggestion
Then('the element with selector {string} should not be empty', (selector) => cy.get(selector).should('not.be.empty'));

```

ndouglas replied:
👍🏻

ndouglas commented on tests/cypress/integration/behavioral/common/i_fill_in_field_with_value.js:
```diff
@@ -15,3 +15,10 @@ Then(`I fill in field with selector {string} with value {string}`, (selector, va
     .type(value, { force: true })
     .blur();
 });
+
+Then(`I fill in autocomplete field with selector {string} with value {string}`, (selector, value) => {
+  cy.get(selector)
+    .focus()
+    .clear({ force: true })
+    .type(value, { force: true })
+});
```
```suggestion
});

```

ndouglas replied:
👍🏻

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+
+  # EVENT FORM SPECS
+
+  Scenario: Confirm that the default time zone when creating an event is set explicitly to Eastern.
+    Given I am logged in as a user with the "content_admin" role
+    And I am at "node/add/event"
+    And the element with selector "#edit-field-datetime-range-timezone-0-timezone" should contain "New York"
```
I'm aware this wasn't originally in the test, but it's good to follow a "Given... When... Then..." pattern.

So in this case it should be 
```suggestion
  Scenario: Confirm that the default time zone when creating an event is set explicitly to Eastern.
    Given I am logged in as a user with the "content_admin" role
    When I am at "node/add/event"
    Then the element with selector "#edit-field-datetime-range-timezone-0-timezone" should contain "New York"
```

Not picking on you, since I (or whoever wrote this test) messed it up from the start... but worth making the improvement while we're in here.  Also, I'll refer to this later, and this is a good clear example.

chri5tia replied:
Ah, good to know! Thank you.

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I select the "At a VA facility" radio button
+    Then I should see "Facility location"
+    And I should see "Building, floor, or room"
+    And I select the "At a non-VA location" radio button
```
```suggestion
    When I select the "At a non-VA location" radio button
```

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I should see "City"
+    And I should see "State"
+    And I should see "Country"
+    And I select the "Online" radio button
```
```suggestion
    When I select the "Online" radio button
```

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    Then I should see an element with the selector "#edit-field-url-of-an-online-event-0-uri"
+
+    # Registration checkbox reveals conditional form elements
+    And I check the "Include registration information" checkbox
```
```suggestion
    When I check the "Include registration information" checkbox
```

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I check the "Include registration information" checkbox
+    Then I should see "Registration is required for this event"
+    And I should see "Call to action"
+    And I select option "Register" from dropdown "Call to action"
+    Then I should see an element with the selector "#edit-field-link-0-uri"
+    And I select option "Apply" from dropdown "Call to action"
+    Then I should see an element with the selector "#edit-field-link-0-uri"
+    And I select option "RSVP" from dropdown "Call to action"
+    Then I should see an element with the selector "#edit-field-link-0-uri"
+    And I select option "More Details" from dropdown "Call to action"
+    Then I should see an element with the selector "#edit-field-link-0-uri"
+    And I select option "- None -" from dropdown "Call to action"
+    Then the element with selector "#edit-field-link-0-uri" should not be visible
```
Generally speaking, the "And ..." should be "When ...".  We should explicitly indicate when we're switching contexts by using **Then** or **When** rather than **And**, which could mean anything.

See #8810.

I generally try to follow a pattern like this:

- **Given**: initial condition, like I'm logged in
- **And**: other initial conditions.  We haven't yet done this, I don't think, but maybe `I have created an "Office" node` would work.  It's something that we want to have done, but isn't under test specifically.
- **When**: the first action we're testing
- **And**: other actions necessary for the test
- **Then**: the first condition we're checking for
- **And**: other conditions we're checking for

The critical thing I think is that after a **Then**, if we're performing another action, we should use **When**.  I don't think we really switch back to **Given**, but we do switch from **Then** -> **When** pretty often.

So instead of 

```
Given I'm logged in
And I have created an "office" node
When I edit the node
And I change the title to "I am the queen of France"
And I click "Save"
Then the page title is "I am the queen of France"
And I edit the node
And I change the title to "I am the king of France"
And I click "Save"
Then the page title is "I am the king of France"
```

we would do:

```
Given I'm logged in
And I have created an "office" node
When I edit the node
And I change the title to "I am the queen of France"
And I click "Save"
Then the page title is "I am the queen of France"
When I edit the node
And I change the title to "I am the king of France"
And I click "Save"
Then the page title is "I am the king of France"
```

It's mostly just a way of indicating that we're performing a new test, or a new step in the test.  There're no programmatic implications that I'm aware of -- it just makes the test a little easier to read and comprehend, I think, and to understand the role of each step in the test.

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I am at "node/add/event"
+
+    # Check registration call to action conditional field
+    And I select option "Register" from dropdown "Call to action"
```
```suggestion
    When I select option "Register" from dropdown "Call to action"
```

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    Then the element with selector "#edit-field-url-of-an-online-event-0-uri" should be empty
+
+    # Check the location type conditional fields
+    And I select the "At a VA facility" radio button
```
```suggestion
    When I select the "At a VA facility" radio button
```

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    And I fill in autocomplete field with selector "#edit-field-url-of-an-online-event-0-uri" with value "https://va.gov"
+    And I select the "At a VA facility" radio button
+    Then the element with selector "#edit-field-facility-location-0-target-id" should be empty
+    And I select the "At a non-VA location" radio button
```
```suggestion
    When I select the "At a non-VA location" radio button
```

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,71 @@ Feature: CMS Users may effectively create & edit content
+    Then the element with selector "#edit-field-address-0-address-address-line1" should be empty
+    And the element with selector "#edit-field-address-0-address-locality" should be empty
+    And the option "- None -" from dropdown "State" should be selected
+    And I select the "Online" radio button
```
```suggestion
    When I select the "Online" radio button
```

ndouglas replied:
This area's a good example of what I mean, since you have 

```
    Then the element with selector "#edit-field-address-0-address-address-line1" should be empty
    And the element with selector "#edit-field-address-0-address-locality" should be empty
    And the option "- None -" from dropdown "State" should be selected
    And I select the "Online" radio button
    Then the element with selector "#edit-field-url-of-an-online-event-0-uri" should be empty
```

The first two "And" steps include a "should be", which indicates that we're testing something from a **Then** context, which is great, but then the third is a direct action and indicates that we've switched back to a **When** context.

ndouglas reviewed (commented):
Looks great!

chri5tia commented:
> > Most likely changes were made to node forms and the tests need to be updated to reflect that.
> 
> No. The tests would prevent that scenario because any pr that changed the node form would not get passing tests so would not get merged. These failures are transient.

@swirtSJW Can you describe "transient" in this context and how I should approach this issue? I checked it against the main branch and got the same results, which makes me think it's out of scope. Thanks for your feedback!

chri5tia commented:
So far, so good. I am not sure why I am getting this. All the other form elements are working and it appears correctly in the browser.
<img width="793" alt="Screen Shot 2022-11-17 at 3 23 19 PM" src="https://user-images.githubusercontent.com/22764938/202582441-17a84416-dafd-4ac5-b477-e17abc5c95f4.png">

ndouglas commented:
Look at the scroll-to-element commands.  It might be that the element isn't visible in the browser view.

chri5tia commented:
Not sure how my branch got it's name, but if I change the name, I would have to cut a new PR, so that's likely not worth it. Let it be known that VACMS-19681 should have been named VACMS-10905...

swirtSJW commented:
> > > Most likely changes were made to node forms and the tests need to be updated to reflect that.
> > 
> > 
> > No. The tests would prevent that scenario because any pr that changed the node form would not get passing tests so would not get merged. These failures are transient.
> 
> @swirtSJW Can you describe "transient" in this context and how I should approach this issue? I checked it against the main branch and got the same results, which makes me think it's out of scope.

Transient means they are hiccups.  Not, as you were proposing, based on changes being made to node forms but not captured at the time by tests.  An changes to the form would be flagged by the test and not allowed to be merged.

swirtSJW reviewed (dismissed):
Very nice work Christia.  Well done.

swirtSJW commented:
Two failures
![image](https://user-images.githubusercontent.com/5752113/202636910-f6d9ba34-a82a-4085-b6b3-ed96c06328dd.png)


![image](https://user-images.githubusercontent.com/5752113/202636873-bdb14255-2acb-4cdb-9da7-54be193ff194.png)


looking into these

swirtSJW commented:
The other is a failure of the test you are adding 
![image](https://user-images.githubusercontent.com/5752113/202638253-bbae6ef4-970c-4480-9c6a-fdb7af62c6d2.png)

swirtSJW commented on tests/cypress/integration/behavioral/common/i_should_see.js:
```diff
@@ -1,19 +1,18 @@
 Then('I should not see xpath {string}', (expression) => cy.xpath(expression).should('not.exist'));
 
-Then('I should see an image with the selector {string}', (selector) => {
+Then('I should see the image with the selector {string}', (selector) => {
```
Truth be told, I like the original language because "I should see an image" implies there could be more than one.   Where "I should see the image" falsely implies there is only one.

swirtSJW commented:
Here is the most recent failure
![image](https://user-images.githubusercontent.com/5752113/202812025-31149bb4-db4e-4a49-815a-3ad4344ea977.png)

swirtSJW commented:
![image](https://user-images.githubusercontent.com/5752113/202812247-b1f4f817-a19a-4d21-8b34-73d7802f48a4.png)

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,102 @@ Feature: CMS Users may effectively create & edit content
+    Then "Facility location" should not be visible
+    And "Building, floor, or room" should be visible
+    And "Street address" should be visible
+    And the element with the selector "edit-field-address-0-address-locality" should be visible
```
It occurred to me that the reason that you couldn't find 'City' was because that was found _first_ in a menu somewhere.  If we can't be more specific, the selector approach should work, but we'd need a `#` or `.`.

swirtSJW replied:
That field id matches what I am seeing.  We'll have to look at the screen grab.

ndouglas replied:
yeah, so if it's an id it'll need a `#`.

swirtSJW replied:
DOh.. I missed reading that in your previous comment.  That would do it.

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,102 @@ Feature: CMS Users may effectively create & edit content
+    Then "Facility location" should be visible
+    And "Building, floor, or room" should be visible
+    And "Street address" should not be visible
+    And the element with the selector "edit-field-address-0-address-locality" should not be visible
```
```suggestion
    And the element with the selector "#edit-field-address-0-address-locality" should not be visible
```

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,102 @@ Feature: CMS Users may effectively create & edit content
+    Then "Facility location" should not be visible
+    And "Building, floor, or room" should be visible
+    And "Street address" should be visible
+    And the element with the selector "edit-field-address-0-address-locality" should be visible
```
```suggestion
    And the element with the selector "#edit-field-address-0-address-locality" should be visible
```

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,102 @@ Feature: CMS Users may effectively create & edit content
+    Then "Facility location" should not be visible
+    And "Building, floor, or room" should not be visible
+    And "Street address" should not be visible
+    And the element with the selector "edit-field-address-0-address-locality" should not be visible
```
```suggestion
    And the element with the selector "#edit-field-address-0-address-locality" should not be visible
```

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,102 @@ Feature: CMS Users may effectively create & edit content
+    #   [va] "Facility location"
+    #   [va][non-va] "Building, floor, or room"
+    #   [non-va] "Street address"
+    #   the element with the selector "edit-field-address-0-address-locality" (City)
```
```suggestion
    #  "#edit-field-address-0-address-locality" (City)
```

swirtSJW commented:
I pushed the fix.  Now we just sit back and watch those sweet tests pass ;)

swirtSJW commented:
his element `<a>` is not visible because its parent `<ul.toolbar-menu.lower-level-nav.menu-level-3>` has CSS property: `display: none`

![image](https://user-images.githubusercontent.com/5752113/202820554-37460448-6618-4fc7-b13e-4960a5bf7790.png)

ndouglas commented:
I think the error there is that it's trying to match text that shows up first, DOM-speaking, in a menu or something else that's hidden.  So you can use an XPath or selector to specify the precise element, but otherwise Cypress is going to match the first DOM element matching that string.

Another possibility is that we could update the step definitions to ignore the toolbar and search for elements within the `<article>` or whatever.  That's probably preferable to using these long, semi-brittle selectors, provided we create an alternate step definition that allows searching the toolbar.

Not 100% that's the issue, but I suspect that it is.

chri5tia commented:
This one is failing because editing the node is locked, and therefor cypress is not able to select the necessary option to show its conditional:

<img width="849" alt="Screen Shot 2022-11-21 at 10 17 18 AM" src="https://user-images.githubusercontent.com/22764938/203130433-d7476d21-1558-4cbe-a34e-ed555fc56a9d.png">

<img width="892" alt="Screen Shot 2022-11-21 at 10 18 32 AM" src="https://user-images.githubusercontent.com/22764938/203130496-934c97af-5307-41b5-8129-79078c1989e6.png">

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -4,7 +4,7 @@ Feature: CMS Users may effectively create & edit content
   I need to have certain functionality available
 
-  Scenario: Log in and confirm that System-wide alerts can be created and edited
+    Scenario: Log in and confirm that System-wide alerts can be created and edited
```
I believe this should be outdented.

chri5tia commented:
@swirtSJW We need to look at the details of the cypress failure and see if it is the one I mentioned above due to a locked node, or if it is different? Ready to re-review.

ndouglas commented:
![Screen Shot 2022-11-21 at 1 52 27 PM](https://user-images.githubusercontent.com/1318579/203136347-7347c8a8-110f-4979-8537-dd9f890f7190.png)

ndouglas commented:
I think it's the Staten Island Vet Center that's causing the issue.  We should probably confine these step definitions to search within `div.page-wrapper` or something like that.

chri5tia commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,102 @@ Feature: CMS Users may effectively create & edit content
+    Then "Facility location" should be visible
+    And "Building, floor, or room" should be visible
+    And "Street address" should not be visible
+    And the element with the selector "#edit-field-address-0-address-locality" should not be visible
```
I had to change it because "city" is in more than one place on the page, so it was passing regardless if the conditional was working... I think. Any suggestions?

ndouglas replied:
K, makes sense.  Carry on 🙂

ndouglas commented on tests/cypress/integration/behavioral/common/should_be_visible.js:
```diff
@@ -0,0 +1,26 @@
+//       See VACMS-11609
+
+Then('the element with the selector {string} should be visible', (selector) => cy.get(selector).should('be.visible'));
+Then('the element with the selector {string} should not be visible', (selector) => cy.get(selector).should('be.visible'));
```
```suggestion
Then('the element with the selector {string} should not be visible', (selector) => cy.get(selector).should('not.be.visible'));
```

chri5tia commented:
Any idea why this is failing?

<img width="330" alt="Screen Shot 2022-11-23 at 12 24 33 PM" src="https://user-images.githubusercontent.com/22764938/203639840-15ea662b-6ea4-4de2-b631-ed7f82cb34d4.png">
<img width="400" alt="Screen Shot 2022-11-23 at 12 24 27 PM" src="https://user-images.githubusercontent.com/22764938/203639850-31daae63-11e0-401a-9143-8f762d8de97e.png">
![CMS Users may effectively create   edit content -- Confirm that the event form conditional elements are shown or hidden appropriately (failed)](https://user-images.githubusercontent.com/22764938/203639904-779ecba1-ec4f-4fb5-923c-b6ccc60c2601.png)

Oh wait, this might refer to the next item, which is

```
    Then I should see the element with the selector "#edit-field-link-0-uri"
    And I select option "More Details" from dropdown "Call to action"
    Then I should see the element with the selector "#edit-field-link-0-uri"
    And I select option "- None -" from dropdown "Call to action"
    Then the element with the selector "#edit-field-link-0-uri" should not be visible
```

Yeah, looks like I confused myself with `with selector` vs. `with the selector`. Looks like I introduced that pattern so I'll remove it and go back to `with selector`.

chri5tia commented:
I decided to go ahead and reset soft through my first commit and instead of trying to chase ghosts and undoing later mistakes. I have removed any consistency and housekeeping changes, which I will save for a different ticket. There should only be one commit and many less files changed. Hoping the behat tests and cypress will pass now.

ndouglas commented on tests/cypress/integration/behavioral/common/the_option_from_dropdown_should_be_selected.js:
```diff
@@ -1,5 +1,9 @@
 import { Given } from "cypress-cucumber-preprocessor/steps";
 
-Given(`the option {string} from dropdown with selector {string} should be selected`, (option, selector) => {
+Given('the option {string} from dropdown {string} should be selected', (option, label) => {
+  cy.findAllByLabelText(label).select(option, { force: true });
+});
+
```
Nice!

ndouglas commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -5,7 +5,7 @@ Feature: CMS Users may effectively create & edit content
 
   Scenario: Log in and confirm that System-wide alerts can be created and edited
-    When I am logged in as a user with the "content_admin" role
+  When I am logged in as a user with the "content_admin" role
```
Think this should be indented two spaces.

swirtSJW replied:
What Nate pointed out still needs to be addressed.
```suggestion
    When I am logged in as a user with the "content_admin" role
```

ndouglas reviewed (dismissed):
This is great work!  I'm sorry it was a painful process.

swirtSJW commented:
Current failing test ────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  behavioral/permissions.feature                                                (12 of 17)


  Permissions
    ✓ Content api consumer can access redirects and graphql api (8453ms)
    ✓ Redirect administrator can add/edit, administer redirects (8392ms)
    ✓ Administer user role can add/edit users (10318ms)
    ✓ Content api consumer cannot alter existing menus (4415ms)
    1) Content Admins should be able to browse sections from their profile even if none have been specifically assigned to them


  4 passing (36s)
  1 failing

  1) Permissions
       Content Admins should be able to browse sections from their profile even if none have been specifically assigned to them:
     TypeError: Cannot read property 'user' of undefined
      at Context.eval (https://pr11613-depqhtkpkw4vjvl1b49c0w230x1rfpxo.ci.cms.va.gov/__cypress/tests?p=tests/cypress/support/index.js:15821:35)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        5                                                                                │
  │ Passing:      4                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     36 seconds                                                                       │
  │ Spec Ran:     behavioral/permissions.feature                                                   │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /var/lib/tugboat/docroot/cypress/screenshots/actual/behavioral/permissions.featu     (1280x720)
     re/Permissions -- Content Admins should be able to browse sections from their pr               
     ofile even if none have been specifically assigned to them (failed).png

ndouglas commented:
That's an odd and unexpected one.  I'll look into it.

ndouglas commented:
![Screen Shot 2022-11-28 at 11 22 33 AM](https://user-images.githubusercontent.com/1318579/204328830-e3e3c21f-b695-4ece-b6ef-c8793e35684c.png)

None of this code was modified, and this function is used in many tests, so I'm guessing this is some kind of glitch.  I'm kinda confused and disturbed by the failure.  I'll re-run it -- I don't think it has any implications for this PR, but it makes me want to live in a cave and eat berries instead of dealing with technology.

swirtSJW reviewed (changes requested):
A couple minor bits

chri5tia commented on tests/cypress/integration/behavioral/common/should_be_visible.js:
```diff
@@ -0,0 +1,26 @@
+//       Then create "Should not exist" to use .should('exist') instead of "I should see"
+//       See VACMS-11609
+
+Then('the element with selector {string} should be visible', (selector) => cy.get(selector).should('be.visible'));
+Then('the element with selector {string} should not be visible', (selector) => cy.get(selector).should('not.be.visible'));
```
For this work, I went with existing patterns. I did a text search of the code and found that `the element` is what was being used and that `an element` was introduced by me. We can save this for a separate ticket.

swirtSJW replied:
"an element" is also an existing pattern that @ndouglas introduced back in april.
![image](https://user-images.githubusercontent.com/5752113/204373673-498e4a0d-4cc9-4e22-996e-d131b79fa575.png)

swirtSJW replied:
> We can save this for a separate ticket. Thanks for your feedback!

As a rule I don't like the pattern of creating a cleanup ticket to clean up new code that is being introduced now.  That is not a good use of our time and cleanup tickets are hard to get prioritized with the client because they offer no gain in functionality. It also creates a bit of fragility, because on refactoring the tests down the road, you will never have the clarity that you have on them now, because now is when you are introducing them.

chri5tia commented:
For visibility, @ndouglas and I decided to change the statements so that "I should see" refers to visibility instead of existence. For the sake of getting this ticket out, we created [a separate ticket ](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11609)to handle that definition change. In that ticket, I can make and test the changes relating to consistent use of `the` vs. `an` and the backtick vs the apostrophe, etc.

chri5tia commented:
I am getting a broken link when trying to look at the cypress test files:
https://tugboat.vfs.va.gov/63751fb94b29f55af683f44e/cypress

ndouglas commented:
you gotta stick that on the end of the `https://pr####` or `https://cms-wpojfwepfojwepofjwefpjofw` URLs, not the Tugboat dashboard URL

chri5tia commented:
Maybe another fluke: my work should not have touched this. Which is why I specifically removed any pattern changes I had introduced in an earlier history.

https://cms-depqhtkpkw4vjvl1b49c0w230x1rfpxo.ci.cms.va.gov/cypress/screenshots/actual/behavioral/file_upload.feature/

![Files can be uploaded and attached to nodes successfully -- We should be able to attach files to event nodes (failed)](https://user-images.githubusercontent.com/22764938/204394748-4c2a80b2-4231-4f0c-974d-979da78afb26.png)

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,94 @@ Feature: CMS Users may effectively create & edit content
+
+    # Registration checkbox reveals conditional form elements
+    When I check the "Include registration information" checkbox
+    Then I "Cost" should be visible
```
```suggestion
    Then "Cost" should be visible
```

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,94 @@ Feature: CMS Users may effectively create & edit content
+    And I select option "- None -" from dropdown "Call to action"
+    Then the element with selector "#edit-field-link-0-uri" should not be visible
+
+    When I uncheck the "Include registration information" checkbox
```
Good. Thanks for adding the  tests for them disappearing.  There is still a minor gap in the test for testing that they were not showing initially but we can call it good enough.

swirtSJW reviewed (changes requested):
Only two required changes.  The others are just food for thought.

chri5tia commented:
> As a rule I don't like the pattern of creating a cleanup ticket to clean up new code that is being introduced now.

For sure! The way I see it *in this instance* is that even though these lines of code are new, they use existing patterns. No where does "an element" exist in the code already. But `the element` does. When I changed it, it created some confusion. Since I already have a housekeeping ticket for some of these things, that's where I am saying I'd like to make this change across the board. I *could* change it for just the lines I added, but it causes confusion I think. This is good food for thought though!

swirtSJW commented:
> . No where does "an element" exist in the code already. 
 
 I already pointed out that, that is not true.  https://github.com/department-of-veterans-affairs/va.gov-cms/pull/11613#discussion_r1034017158

There is an existing pattern for both.  I am trying to suggest that we use the more honest pattern.  It doesn't have to hold up this PR.

swirtSJW reviewed (dismissed):
Approved.   I just rebased to trigger the test run again.  I think Nate's fix elswhere may resolve the failing test.

swirtSJW commented:
The failing test if coming from this PR
```
Running:  behavioral/content_editing.feature                                             (1 of 17)


  CMS Users may effectively create & edit content
    ✓ Log in and confirm that System-wide alerts can be created and edited (48078ms)
    ✓ Confirm that content cannot be published directly from the node view but can from the node edit form. (21298ms)
    ✓ Confirm Generate automatic URL alias is unchecked after node publish. (31977ms)
    ✓ Confirm Generate automatic URL alias is unchecked after taxonomy term publish. (14479ms)
    ✓ Confirm that press release country fields are shown correctly (11131ms)
    ✓ Log in, edit, and save nodes with save and continue button and confirm revision saves changes. (31035ms)
    ✓ Confirm that the default time zone when creating an event is set explicitly to Eastern. (10165ms)
    1) Confirm that the event form conditional elements are shown or hidden appropriately
    ✓ Confirm that event form conditional fields are cleared out if parent options change (19621ms)


  8 passing (3m)
  1 failing

  1) CMS Users may effectively create & edit content
       Confirm that the event form conditional elements are shown or hidden appropriately:
     Error: Step implementation missing for: "Registration is required for this event"  should be visible
      at Context.resolveAndRunStepDefinition (https://pr11613-depqhtkpkw4vjvl1b49c0w230x1rfpxo.ci.cms.va.gov/__cypress/tests?p=tests/cypress/integration/behavioral/content_editing.feature:27195:11)
      at Context.eval (https://pr11613-depqhtkpkw4vjvl1b49c0w230x1rfpxo.ci.cms.va.gov/__cypress/tests?p=tests/cypress/integration/behavioral/content_editing.feature:26510:35)




  (Results)

  ┌─────────────────���──────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        9                                                                                │
  │ Passing:      8                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     3 minutes, 24 seconds                                                            │
  │ Spec Ran:     behavioral/content_editing.feature                                               │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /var/lib/tugboat/docroot/cypress/screenshots/actual/behavioral/content_editing.f     (1280x720)
     eature/CMS Users may effectively create & edit content -- Confirm that the event               
      form conditional elements are shown or hidden appropriately (failed).png                      

```

swirtSJW commented on tests/cypress/integration/behavioral/content_editing.feature:
```diff
@@ -93,3 +88,93 @@ Feature: CMS Users may effectively create & edit content
+    # Registration checkbox reveals conditional form elements
+    When I check the "Include registration information" checkbox
+    Then "Cost" should be visible
+    And "Registration is required for this event"  should be visible
```
```suggestion
    And "Registration is required for this event" should be visible
```

swirtSJW commented:
Build failed on tugboat.  I just triggered a rebuild.

chri5tia commented:
There were some more issues with me mixing up `an ... with the selector` vs `an ... with selector` vs `the ... with the selector` vs `the ... with selector` statements, so I went ahead and changed the ones I added to `an ... with the selector`.

swirtSJW reviewed (approved):
Looks great and tests passed Woot!

swirtSJW commented:
![image](https://user-images.githubusercontent.com/5752113/205217681-ce6f96ed-dfbf-4790-9067-84c3c35d63ee.png)
